mempool/src/pool/store/mem_usage.rs

```diff
@@ -0,0 +1,443 @@
+        let leaves = (leaf_size * num_elems) / CAP as u64;
+
+        // Size of internal nodes. We add extra 10% overhead for all the levels of the tree
+        let elems_per_internal_node = (CAP * BF) as u64;
+        let internals = (internal_size * num_elems * 11) / (elems_per_internal_node * 10);
+
+        // Total size of the B-tree structure. Assuming nodes are on average 75% full,
+        // an additional overhead is added for the unused occupied space.
+        let total = 4 * (leaves + internals) / 3;
+
+        total as usize
+    }
+}
+
+/// Trait for data types capable of reporting their current memory usage
+///
+/// TODO: Make this a derivable trait so the `impl`s react automatically to changes.
+pub trait MemoryUsage {
+    /// Get amount of memory taken by the data owned by `self` (e.g. if it contains `Box` or `Vec`)
+    fn indirect_memory_usage(&self) -> usize;
+}
+
+/// Total memory usage (indirectly by pointers + for the object itself)
+fn total_memory_usage<T: MemoryUsage>(val: &T) -> usize {
+    val.indirect_memory_usage() + mem::size_of::<T>()
+}
+
+macro_rules! impl_no_indirect_memory_usage {
+    ($($ty:ty),* $(,)?) => {
+        $(
+            impl MemoryUsage for $ty {
+                fn indirect_memory_usage(&self) -> usize { 0 }
+            }
+        )*
+    };
+}
+
+impl_no_indirect_memory_usage!((), bool, usize, u8, u16, u32, u64, u128);
+
+impl<K, V> MemoryUsage for std::collections::BTreeMap<K, V> {
+    /// The mem usage for [BTreeMap].
+    ///
+    /// Includes the nodes and the key and value data stored in the nodes. It does not, however,
+    /// include the memory taken up by data keys and values point to indirectly. Any indirect data
+    /// has to be tracked separately. This is so that the memory usage of the B-tree map/set can be
+    /// calculated from the number of elements alone without any expensive traversals.
+    fn indirect_memory_usage(&self) -> usize {
+        btree::usage::<K, V>(self.len())
+    }
+}
+
+impl<K> MemoryUsage for std::collections::BTreeSet<K> {
+    /// Same limitation as for `BTreeMap` also applies here
+    fn indirect_memory_usage(&self) -> usize {
+        btree::usage::<K, ()>(self.len())
+    }
+}
+
+impl<T: MemoryUsage> MemoryUsage for Option<T> {
+    fn indirect_memory_usage(&self) -> usize {
+        self.as_ref().map_or(0, |x| x.indirect_memory_usage())
+    }
+}
+
+impl<T: MemoryUsage> MemoryUsage for &[T] {
+    fn indirect_memory_usage(&self) -> usize {
+        self.iter().map(T::indirect_memory_usage).sum::<usize>() + mem::size_of_val::<[T]>(*self)
+    }
+}
+
+impl<T: MemoryUsage> MemoryUsage for Vec<T> {
+    fn indirect_memory_usage(&self) -> usize {
+        self.as_slice().indirect_memory_usage()
+    }
+}
+
+impl<T: MemoryUsage> MemoryUsage for Box<T> {
+    fn indirect_memory_usage(&self) -> usize {
+        total_memory_usage::<T>(self.as_ref())
+    }
+}
+
+impl<T> MemoryUsage for common::primitives::Id<T> {
```

Comment:
use `impl_no_indirect_memory_usage` for this?

Reply:
It's a stupidly simple macro that does not handle generics. Not sure it's worth adding the support for generics just for this one thing.

Reply:
ah ok nevermind then